daemon/restart.go

```diff
@@ -54,12 +54,23 @@ func (daemon *Daemon) containerRestart(ctx context.Context, container *container
 	if container.IsRunning() {
 		// set AutoRemove flag to false before stop so the container won't be
 		// removed during restart process
-		autoRemove := container.HostConfig.AutoRemove
-
+		container.Lock()
+		restoreAutoRemove := container.HostConfig.AutoRemove
 		container.HostConfig.AutoRemove = false
+		container.Unlock()
+
 		err := daemon.containerStop(ctx, container, options)
-		// restore AutoRemove irrespective of whether the stop worked or not
-		container.HostConfig.AutoRemove = autoRemove
+
+		// Restore AutoRemove irrespective of whether the stop worked or not.
+		// Do it after the container has been started (or the stop failed), so it doesnt'
+		// get restored to true BEFORE the auto removal is actually considered by the monitor.
+		if restoreAutoRemove {
+			defer func() {
```

Comment:
If we call this in a `defer`, it will be called after `daemon.checkpointAndSave`, which I think will mean that the container configuration (on disk) would now always have `AutoRemove: false` 🤔

Reply:
That said; I wonder if we could use the `Restarting` state of the container somehow; there already seems to be some code for that, so it's a bit confusing why we need to also set the auto-remove option; https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/monitor.go#L65-L71

Reply:
The `err == nil` in the first branch of the above looks somewhat fishy; it would mean that if an error occurs, we'll go ahead and stop (and remove) the container; perhaps that's ok, but wondering what errors can occur.

There's also a `daemon.autoRemove()` further down in that same function (logic is confusing and complicated)

Reply:
The err here is not container error, but error returned from `ShouldRestart`. err is not nil for example when restart is cancelled for some reason, for example the container with restart policy was stopped manually.
I included some more logging here and moved handling the err a bit closer to the ShouldRestart call.